### ISSUE TITLE
'line-height: normal' definition should reflect reality of determining based on fonts

### ISSUE THREAD

**Issue by dbaron** [w3c/csswg-drafts#1254]
The [definition of `line-height: normal`](https://drafts.csswg.org/css2/visudet.html#propdef-line-height) says:
> Tells user agents to set the used value to a "reasonable" value based on the font of the element. The value has the same meaning as &lt;number>. We recommend a used value for 'normal' between 1.0 to 1.2. The computed value is 'normal'. 

This needs to explain what happens when font fallback occurs and multiple fonts are used.  Implementations use font metrics to get the normal line-height (at least some use external leading or line gap), and these metrics can be different in different fonts.  There may also be an interesting interaction with the [definition of the content height](https://drafts.csswg.org/css2/visudet.html#inline-non-replaced), which separately considers font fallback.

More research is needed here as to what implementations do and what they should do, in particular:
* what font metrics are used
* how the metrics from multiple fonts are combined, and whether this has an interaction with the computation of the content box height or whether it occurs entirely separately

/cc @litherum

**Comment by dbaron** [comment]
This is related to the [discussion that the working group is having right now](https://log.csswg.org/irc.w3.org/css/2017-04-19/#e797903).

**Comment by dbaron** [comment]
I'd also note that CSS2 10.8.1 says:
> Still for each glyph, determine the leading L to add, where L = 'line-height' - AD. Half the leading is added above A and the other half below D, giving the glyph and its leading a total height above the baseline of A' = A + L/2 and a total depth of D' = D + L/2.
> The height of the inline box encloses all glyphs and their half-leading on each side and is thus exactly 'line-height'. Boxes of child elements do not influence this height.

This model seems to suggest that if we add external leading from fonts for normal line height, we should probably do so separately for each glyph, i.e., using a per-glyph `line-height` when computing L in the algorithm above.


Gecko's implementation has a number of bugs [described in bug 1358377](https://bugzilla.mozilla.org/show_bug.cgi?id=1358377), so I'm not sure it's particularly useful in determining what to do here.  Information from other implementations may be more useful.

**Comment by kojiishi** [comment]
Blink does for `line-height: normal`:

1. Collect used fonts for the run.
2. For each used fonts:
  1. Compute A and D from the font metrics (note, IIUC, our A/D are slightly different from Gecko's)
  2. Compute L/2 using line spacing from the font metrics ((line spacing - (A + D)) / 2)
3. Compute the union of A' (A + L/2) and D' (D + L/2) of all used fonts.

I believe this is mostly the same as WebKit.

**Comment by kojiishi** [comment]
Note, the "used fonts" include system fallbacks (i.e., that are not in `font-family`.)

**Comment by astearns** [comment]
https://lists.w3.org/Archives/Public/www-archive/2017Apr/att-0008/IMG_20170421_173959.jpg

**Comment by dbaron** [comment]
For what it's worth, by the end of the discussion above in https://github.com/w3c/csswg-drafts/issues/1254#issuecomment-296129092, I'm not sure we still agreed with the first of the two resolutions, although we hadn't resolved on something to replace it.  (It was perhaps a little prematurely recorded as a resolution?)

**Comment by frivoal** [comment]
Right. I believe that `RESOLVED: add leading to union of ascent and descent` was premature, and what we now need is to check whether the various implementations are doing the blue, black, or red thing in [the drawing](https://lists.w3.org/Archives/Public/www-archive/2017Apr/att-0008/IMG_20170421_173959.jpg), then assuming they're not all doing the same thing, have another round of discussions to figure out which one we can agree to align on.

@astearns @atanassov, do you want to distribute actions to the various browser vendors to go check that?

**Comment by fantasai** [comment]
[Minutes of the 2017 Tokyo F2F discussion](https://lists.w3.org/Archives/Public/www-style/2017May/0055.html)
[Minutes of 2017 Paris F2F discussion](https://lists.w3.org/Archives/Public/www-style/2017Sep/0003.html)
[Minutes of 2017 Sept 27 telecon](https://lists.w3.org/Archives/Public/www-style/2017Sep/0054.html)
[Minutes of 2017 SF F2F discussion = resolutions on definitions of line-height calculations](https://lists.w3.org/Archives/Public/www-style/2017Dec/0007.html)

[Meta-bug on line-height issues](https://github.com/w3c/csswg-drafts/issues/1796)

I believe most of this is captured in the current CSS Inline Layout draft in this section, which derives from CSS2.1&sect;10.8: https://drafts.csswg.org/css-inline-3/#inline-height

What wasn't covered there was how the font's own line-gap/line-height metrics are handled. It's clear from the discussion that when they apply, the gap is split in half, and half applied above and half below the baseline. What's not entirely clear to me is if this is done:
- only when computing the height of the line box
- also when establishing the inline’s content edges (for background painting)
- also when aligning by `vertical-align: text-top`

I suspect we only want the first one, so I've gone ahead and made that explicit.

Agenda+ to review this all with the CSSWG.
